tests/isa-test.cc

```diff
@@ -0,0 +1,774 @@
+	 * Optionally assert that we receive CHERI fault `expectedFault`. The fault
+	 * is optional because certain faults should not occur if the stored
+	 * capability is untagged. Because we are storing a local capability into a
+	 * global it can only succeed if the stored capability has the tag cleared.
+	 */
+	void do_store_test(std::optional<CauseCode> expectedFault,
+	                   CapabilityFilter<int *>  baseFilter = identity_filter,
+	                   CapabilityFilter<int>    dataFilter = identity_filter,
+	                   std::optional<size_t>    anExpectedMCause = MCauseCheri)
+	{
+		int        myLocalInt      = 0;
+		Capability capToIntPointer = baseFilter(myGlobalIntPointer);
+		Capability storeData       = dataFilter(&myLocalInt);
+		bool       expectCrash     = anExpectedMCause.has_value();
+		if (expectCrash)
+		{
+			expectedMCause    = anExpectedMCause;
+			expectedCauseCode = expectedFault;
+			expectedErrorPC   = {address_of_pointer(faulting_store)};
+		}
+		int previousCrashes = crashes;
+		faulting_store(storeData, capToIntPointer);
+		TEST(crashes == previousCrashes + (expectCrash ? 1 : 0),
+		     "{} store of {} via {} to crash",
+		     expectCrash ? "Expected" : "Did not expect",
+		     storeData,
+		     capToIntPointer);
+	}
+
+	void test_store_faults()
+	{
+		debug_log("Attempting to store via untagged capability");
+		do_store_test(CauseCode::TagViolation, clear_tag_filter);
+
+		debug_log("Attempting to store via sealed capability");
+		do_store_test(CauseCode::SealViolation, get_sealed_capability_filter);
+
+		debug_log("Attempting to store via read-only capability");
+		do_store_test(CauseCode::PermitStoreViolation,
+		              permission_filter<PermissionSet{Permission::Load}>);
+
+		debug_log("Attempting to store capability via data only pointer");
+		do_store_test(CauseCode::PermitStoreCapabilityViolation,
+		              permission_filter<PermissionSet{Permission::Load,
+		                                              Permission::Store}>);
+
+		debug_log("Attempting to store local capability in global");
+		do_store_test(
+		  CauseCode::PermitStoreLocalCapabilityViolation,
+		  permission_filter<PermissionSet{Permission::Load,
+		                                  Permission::Store,
+		                                  Permission::LoadStoreCapability}>);
+
+		debug_log("Attempt to store capability via too small bounds");
+		do_store_test(
+		  CauseCode::BoundsViolation, restrict_bounds_filter, clear_tag_filter);
+
+		debug_log("Attempt to store capability via misaligned pointer");
+		do_store_test(std::nullopt,
+		              misalign_filter,
+		              clear_tag_filter,
+		              priv::MCAUSE_STORE_MISALIGNED);
+
+		debug_log(
+		  "Attempting to store untagged capability via data only pointer");
+		// Storing untagged cap via data-only cap is allowed.
+		do_store_test(
+		  std::nullopt,
+		  permission_filter<PermissionSet{Permission::Load, Permission::Store}>,
```

Comment:
Why not `identity_filter`?

Reply:
Because the intent of the this is to check that storing an untagged capability through a capability without MC does not fault (whereas storing a tagged capability would fault, and this is tested previously). Therefore we have to strip MC from the authority cap.